File: lib/wDispatcher/parentreporter.cpp
#include "parentreporter.h"

W::ParentReporter::ParentReporter():
    W::DefaultHandler(),
    handlers()
{
}

W::ParentReporter::~ParentReporter()
{
}

bool W::ParentReporter::call(const W::Event& ev) const
{
    const W::Address& addr = ev.getDestination();
    std::map<int, W::Handler*> result;
    
    Hmap::const_iterator itr = handlers.begin();
    Hmap::const_iterator end = handlers.end();
    
    for (; itr != end; ++itr) {                 //need to find the closest parent to the event destination
        if (addr.begins(itr->first)) {          //the closest parent has the longest address of those whose destinatiion begins with
            result.insert(std::make_pair(itr->first.size(), itr->second));
        }
    }
    if (result.size() > 0) {
        std::map<int, W::Handler*>::const_iterator itr = result.end();
        --itr;
        itr->second->pass(ev);                  //need to report only to the closest parent
        return true;
    } else {
        return false;
    }
}

void W::ParentReporter::registerParent(const W::Address& address, W::Handler* handler)
{
    Hmap::const_iterator itr = handlers.find(address);
    if (itr != handlers.end()) {
        throw 1;
    } else {
        handlers.insert(std::make_pair(address, handler));
    }
}

File: libjs/wDispatcher/parentreporter.js
"use strict";

var DefaultHandler = require("./defaulthandler");
var Handler = require("./handler");

var Address = require("../wType/address");
var AbstractMap = require("../wContainer/abstractmap");

var HandlersMap = AbstractMap.template(Address, Handler);

var ParentReporter = DefaultHandler.inherit({
    "className": "ParentReporter",
    "constructor": function() {
        DefaultHandler.fn.constructor.call(this);
        
        this._handlers = new HandlersMap(false);
    },
    "destructor": function() {
        this._handlers.destructor();
        
        DefaultHandler.fn.destructor.call(this);
    },
    "call": function(ev) {
        var addr = ev.getDestination();
        var result = [];
        
        var itr = this._handlers.begin();
        var end = this._handlers.end();
        for (; !itr["=="](end); itr["++"]()) {
            var pair = itr["*"]();
            if (addr.begins(pair.first)) {
                result[pair.first.size()] = pair.second;        //it's a dirty javascript trick
            }                                                   //I need the longest of matching, and that's how I'm gonna get it
        }
        if (result.length) {
            result[result.length - 1].pass(ev);
            return true;
        } else {
            return false;
        }
    },
    "registerParent": function(parentAddr, handler) {
        this._handlers.insert(parentAddr, handler);
    },
    "unregisterParent": function(parentAddr) {
        var itr = this._handlers.find(parentAddr);
        if (!itr["=="](this._handlers.end())) {
            this._handlers.erase(itr);
        } else {
            throw new Error("An attempt to unregister unregistered parent in ParentReporter");
        }
    }
});

module.exports = ParentReporter;
